### PR TITLE
Autodiscover provider for Nomad

### PR DESCRIPTION
**Pull request by jorgelbg**
At trivago we run an internal cloud using Nomad from Hashicorp. Our logging solution is based on ELK and we use Filebeat to ship the logs from our client nodes into Kafka where it is later on ingested into Elasticsearch using Logstash. Previously we used the and input looking for new jobs in a defined `path`, but the logs lacked a lot of context/metadata from the Job definition/allocation.

This PR adds a new discover module (architecture based on the Kubernetes module). With this new provider, it is possible to start new harvesters by looking at the jobs allocated on each node. We currently run filebeat as a system job on each node and each filebeat instance is responsible for enriching and shipping the local logs.

Example of the configuration for the new provider:

```yaml
filebeat.autodiscover:
  providers:
    - type: nomad
      host: {{ env "node.unique.name" }}
      hints.enabled: true
      hints.default_config:
        type: log
        paths:
          - /appdata/nomad/alloc/${data.meta.uuid}/alloc/logs/*stderr.[0-9]*
          - /appdata/nomad/alloc/${data.meta.uuid}/alloc/logs/*stdout.[0-9]*
```

By using the autodiscover module it is possible to define custom processors using the `meta` stanza on the Nomad job (similar to how it is defined using labels on Kubernetes). For instance:

```
task "nginx-web" {
    driver = "docker"

    meta {
    task-key = "custom-meta"
    "co.elastic.logs/processors.dissect.tokenizer" = "%{ip} - %{user} [%{local_time}] \"%{request}\" %{status} %{bytes_sent} \"%{referer}\" \"%{user_agent}\""
    }
}
```
This example defines a custom `dissect` tokenizer for the logs of this specific task that adds the `dissect` field with a content similar to:

```json
"dissect": {
    "bytes_sent": "7231",
    "referer": "http://nginx-web.prod.trivago.com/",
    "user_agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_14_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/78.0.3904.97 Safari/537.36",
    "ip": "10.2.10.138",
    "user": "-",
    "local_time": "15/Nov/2019:09:04:04 +0000",
    "request": "GET / HTTP/1.1",
    "status": "200"
}
```

By default the following fields are added from the Nomad job/allocation:
* `job`
* `namespace`
* `status`
* `type` (job type: system/service/batch)
* `task.*` (information about the task and custom metadata defined in the job/group/task using the `meta` stanza)
* `datacenters`
* `region`

The PR also includes an `add_nomad_metadata` processor that matches events to specific allocations and adds the metadata.

We've been running this in our production clusters for a few weeks now.

#### TODO:
- [ ] Metricbeat support for extracting stats from the Nomad allocation
- [x] Documentation (it is a WIP and I will add a new commit with the documentation)
- [x] CHANGELOG
- [x] Fields reference

### CODE REVIEW

**Comment by sorantis** [comment]
Thanks for filing the nomad provider, @jorgelbg! This PR is scheduled for review.
We're adding new providers under the x-pack directory. Would you mind moving this provider to `x-pack/libbeat/autodiscover/providers/`?

**Comment by jorgelbg** [comment]
@sorantis This PR includes a new processor (`add_nomad_metadata`) that was implemented (in the current PR) as a libbeat processor, with a custom matcher for filebeat. Should I move those as well to `./x-pack/libbeat/processors/` and `./x-pack/filebeat/processors/`?

**Comment by sorantis** [comment]
@jorgelbg yes, the processor should be moved to `/x-pack`.

**Comment by jorgelbg** [comment]
A couple of tests are failing in travis but it doesn't seem related to the PR itself. One is failing due to:

```
fatal error: runtime: out of memory
```

the other one failed because of a `401` from ES:

```
--- FAIL: TestLoadBeatsTemplate_fromFile (0.00s)
    testing.go:52: 401 Unauthorized: {"error":{"root_cause":[{"type":"security_exception","reason":"unable to authenticate user [beats] for REST request [/]","header":{"WWW-Authenticate":"Basic realm=\"security\" charset=\"UTF-8\""}}],"type":"security_exception","reason":"unable to authenticate user [beats] for REST request [/]","header":{"WWW-Authenticate":"Basic realm=\"security\" charset=\"UTF-8\""}},"status":401}
```

File: libbeat/docs/shared-autodiscover.asciidoc
Comment (jsoriano) [inline libbeat/docs/shared-autodiscover.asciidoc]: Nomad fields will need to be documented. Could you please add a `fields.yml` file in `x-pack/libbeat/processors/add_nomad_metadata/_meta/` with the fields definitions?
You can see some examples in `x-pack/libbeat/processors/add_cloudfoundry_metadata/_meta/fields.yml` or `libbeat/processors/add_kubernetes_metadata/_meta/fields.yml`.

File: x-pack/libbeat/autodiscover/providers/nomad/nomad.go
Comment (jorgelbg) [reply x-pack/libbeat/autodiscover/providers/nomad/nomad.go]: Good point! 
Perhaps we could avoid adding 2 keys by renaming `meta` to `nomad`? In that regard, it would be similar to the kubernetes provider, something like: `${data.meta.alloc_id}` would become `${data.nomad.alloc_id}`.

Comment (jsoriano) [reply x-pack/libbeat/autodiscover/providers/nomad/nomad.go]: For convenience we would need to add both keys. The first one would be enough to have `${data.nomad.alloc_id}`, the second one would be used to enrich events. Fields under `"meta"` are the ones to all events.
If we only add the first one no enrichment would be made to events, and if we only add the second one we would still need to use `meta` namespace in templates (what would become `${data.meta.nomad.alloc_id}`.

File: x-pack/libbeat/common/nomad/watcher.go
Comment (jorgelbg) [reply x-pack/libbeat/common/nomad/watcher.go]: We initially considered the option of _streaming_ the logs instead of running filebeat in each host, which meant that it could potentially run on a host without an agent, but given the current implementation, this sounds better.

Comment (jsoriano) [reply x-pack/libbeat/common/nomad/watcher.go:78]: Yes, I agree with keeping an option to configure this for scenarios like the one you mention where the agent is not running or is not reachable. But taking the id from the host in the agent, if available, looks like a nice default.

File: x-pack/libbeat/common/nomad/metadata.go
Comment (jorgelbg) [reply x-pack/libbeat/common/nomad/metadata.go:72]: For us it has been working fine, it depends if the job (and allocations) are successfully running. Did the Nomad UI/CLI report a different status for the job?

File: x-pack/libbeat/common/nomad/metadata.go
Comment (jorgelbg) [reply x-pack/libbeat/common/nomad/metadata.go:68]: The decision, in this case, was to match the [nomad job stanza](https://www.nomadproject.io/docs/job-specification/job#datacenters) but it is easy to change/fix so I could do it anyhow.

File: x-pack/libbeat/common/nomad/util.go
Comment (jorgelbg) [reply x-pack/libbeat/common/nomad/util.go:16]: The documentation is definitively out of sync. In this case, I'm not really sure what would be best. The nomad API client by default reads several config options (the same that we would set through the custom configuration) from environment variables (https://github.com/hashicorp/nomad/blob/master/api/api.go#L234-L289), which has proven to be enough for our use case.

In the [kubernetes case I see](https://github.com/elastic/beats/blob/2a643c9b7d06e4d6f4d7f3d95493b4efd0c53ac0/libbeat/common/kubernetes/util.go#L46-L62) that configuration options can be set through environment variables or from kubeconfig.

Not sure what is the preferred way from the beats specific case.

**Comment by jsoriano** [comment]
@jorgelbg thanks for updating the PR! I will take another look hopefuly today or tomorrow.

**Comment by jsoriano** [comment]
jenkins run the tests please

File: x-pack/libbeat/common/nomad/metadata.go
Comment (jsoriano) [reply x-pack/libbeat/common/nomad/metadata.go:72]: I will try again. I only tested on a dev agent. But it sounds good if it works well for you on a real cluster.

File: x-pack/libbeat/common/nomad/metadata.go
Comment (jsoriano) [reply x-pack/libbeat/common/nomad/metadata.go:68]: Ok, let's leave it as is.

File: x-pack/libbeat/common/nomad/util.go
Comment (jsoriano) [reply x-pack/libbeat/common/nomad/util.go]: We prefer to expose the main configuration options in the configuration file. Having support for environment variables in addition to that is also ok. But we can revisit this in the future, lets go with this by now.

**Comment by sorantis** [comment]
any progress on this?

**Comment by jsoriano** [comment]
@jorgelbg will you have some time to continue with this PR? If not, it is ok, we will reserve some time to complete it. Thanks for your effort so far!

**Comment by jorgelbg** [comment]
@jsoriano @sorantis Sorry for the delay, I will fix the conflicts and go over the notes from the last review and push the new changes.

**Comment by ddreier** [comment]
Just want to throw out that we're a pretty large Nomad user as well, and are super excited to see this get added. Planning on starting some testing soon. Thanks!

**Comment by jsoriano** [comment]
jenkins run the tests please

**Comment by jorgelbg** [comment]
Small update: applied the changes for the hints-based auto-discover to use `data.nomad` instead of `data.meta`, and updated the documentation.

> 
> * I still have `pending` status in all events, no idea why. If it works for you, maybe this is something on my dev agent.

I will run a test spinning up a local nomad agent with a test job to verify that it is not an oversight (or version related), so far I've been testing against our production clusters.

> * When enabling debug, I see the provider generating stop events for dead allocations. In principle this is not causing any problem, so we can revisit this later. The events I see are like this one:
> 
> ```
> 2020-07-30T12:51:48.390+0200	DEBUG	[autodiscover]	autodiscover/autodiscover.go:236	Got a stop event: map[config:[0xc000c97c20] host:voyager id:410d140e-3e9c-0db6-5f77-fffdd34cc2d8-consul-dev meta:{"nomad":{"alloc_id":"410d140e-3e9c-0db6-5f77-fffdd34cc2d8","datacenters":["dc1"],"job":"consul","name":"consul.server[0]","namespace":"default","region":"global","status":"pending","task":{"name":"consul-dev","service":{"canary_tags":[],"name":[],"tags":[]}},"type":"service"}} nomad:{"alloc_id":"410d140e-3e9c-0db6-5f77-fffdd34cc2d8","datacenters":["dc1"],"job":"consul","name":"consul.server[0]","namespace":"default","region":"global","status":"pending","type":"service"} provider:b893019a-fc2e-4eb8-809b-82585e1521c8 stop:true]
> ```

Usually, the dead allocations are kept around until they're GCed, I don't think that we actively filter these on the very initial sync but it should be possible to implement, it really doesn't make a lot of sense to generate these stop events :).

**Comment by jorgelbg** [comment]
@jsoriano Regarding the difference between the status reported by the nomad CLI and filebeat, I was able to reproduce it, even in our production cluster. Not sure _why_ it happens but when we query the job status via the allocation:

```go
alloc, _, err := client.Allocations().Info(allocID, &qop)
status := *alloc.Job.Status
```

it sometimes returns `pending` when the job is successfully allocated and running. If we query the jobs API directly:

```go
jobs, _, err := client.Jobs().PrefixList("perf-filebeat-demo")
status := jobs[0].Status
```

We get the expected `running` status. This is what the Nomad CLI does: https://github.com/hashicorp/nomad/blob/master/command/job_status.go#L148.

Strangely, it doesn't happen for all allocation/jobs.

**Comment by jsoriano** [comment]
@jorgelbg thanks for the investigation on the status reported. I can also confirm that the allocations and jobs apis return different values:
```
$ curl -s 127.0.0.1:4646/v1/allocation/859e12ae-559d-72f4-8db9-0f0c8d2d088c | jq '.Job.Name, .Job.Status, .Job.ID'
"consul"
"pending"
"consul"
```
```
$ curl -s 127.0.0.1:4646/v1/job/consul | jq '.Name, .Status, .ID'
"consul"
"running"
"consul"
```

This might be a bug in nomad.

WDYT about using the job API to get this status? Other option is to don't get this status by now so we don't block this PR because of this, and we can add it as a follow up.

**Review by jsoriano — COMMENTED** [review]
A couple of small comments about the configuration.

**Comment by jsoriano** [comment]
@jorgelbg I have opened an issue in Nomad project about the incoherences we have found https://github.com/hashicorp/nomad/issues/9386

**Comment by jorgelbg** [comment]
> WDYT about using the job API to get this status? Other option is to don't get this status by now so we don't block this PR because of this, and we can add it as a follow up.

Yep, I was thinking that we could instead rely on `ClientStatus` for this `status` field, which refers to the status of the current allocation in the given Nomad client, AFAIK. Especially considering that the log events that we generate are at the allocation/task level. It is possible that one allocation has already the `running` status but a different allocation is still `pending` or `failed`.

**Comment by jstenvall-ducksify** [comment]
We have a use case where a single application can output more than one logfile. Would this PR be able to handle this case?

From what I can see the is that one pattern will be applied to all log files specified in the filebeat configuration:

     task "nginx-web" {
        driver = "docker"

        meta {
            task-key = "custom-meta"
            "co.elastic.logs/processors.dissect.tokenizer" = "%{ip} - %{user} [%{local_time}] \"%{request}\" %{status} %{bytes_sent} \"%{referer}\" \"%{user_agent}\""
        }
    }

Will there be a possibility to apply a dissect, let's say, per log file ?

**Comment by jsoriano** [comment]
@jstenvall-ducksify thanks for your comment, this is something definitely to take into account, but if this is not supported by this PR, I would suggest to add this in a future change.

We still have to complete nomad support, but for example in kubernetes autodiscover it is possible to define [multiple sets of hints](https://www.elastic.co/guide/en/beats/filebeat/7.10/configuration-autodiscover-hints.html#_multiple_sets_of_hints), that produce multiple sets of configurations. This could be used to define a configuration for each file in case a task logs to multiple files.

In see in your case that you are using the docker driver for the nomad task. Are the multiple log files mounted in some volume or the multiple log files are stdout and stderr? If they are stderr and stdout, there are ways to configure filebeat modules to consume  each one of the streams with different filesets.

**Comment by jstenvall-ducksify** [comment]
> @jstenvall-ducksify thanks for your comment, this is something definitely to take into account, but if this is not supported by this PR, I would suggest to add this in a future change.

Ok, thats good news it's on the radar.

> I see that in your case you are using the docker driver for the nomad task. Are the multiple log files mounted in some volume or the multiple log files are stdout and stderr? If they are stderr and stdout, there are ways to configure filebeat modules to consume each one of the streams with different filesets.

I actually copy/pasted the example from above. My case is a bit different since I have to support legacy java applications (raw_exec) that produces one or multiple log files for specific business reasons. The log files are located in a mounted volume for posterity (manual tail/grep actions). Don't flame me for this, it's like that :).

So, I would like to have the possibility to specify a "pattern" per log file. I have no idea about the effort needed, I'm simply stating my need while trying to be humble (and grateful for all your work).

**Comment by jsoriano** [comment]
Hey @jorgelbg,

I am giving a try to address the open comments in https://github.com/elastic/beats/pull/23324, feel free to cherry-pick these changes in your PR if you want and can continue with this. If not, no problem, I will try to get this ready to be merged after new year. After merging it we can incrementally add further fixes or enhancements.

Thanks!

**Comment by jsoriano** [comment]
jenkins run the tests please

**Review by jsoriano — APPROVED** [review]
Thanks @jorgelbg!

**Comment by jorgelbg** [comment]
@jsoriano I cherry-picked the changes from your branch, but I'm ok if we merge from any side. Thanks for taking care of fixing those issues (especially the Windows test). Changes look great! I was planning on jumping into this after the holidays but glad that you were faster!

**Comment by jsoriano** [comment]
@jorgelbg could you please update the branch with master? Failing tests were fixed yesterday.

**Comment by jsoriano** [comment]
jenkins run the tests please

**Comment by jsoriano** [comment]
Merged, thanks a lot @jorgelbg!

**Comment by sorantis** [comment]
Thank you all for working on this!

**Pull request merged**
